File: src/com/venus/Tree/AVLTree.java
package com.venus.Tree;

import java.util.Comparator;

@SuppressWarnings("ALL")
public class AVLTree<E> extends BBSTree<E> {

    public AVLTree() {
        this(null);
    }

    public AVLTree(Comparator<E> comparator) {
        super(comparator);
    }

    @Override
    public void afterAdd(Node<E> node) {

        while ((node = node.parent) != null) {
            if (isBalanced(node)) {
                updateHeight(node);
            }else {

                rebalance(node);
                break;
            }
        }
    }

    @Override
    public void afterRemove(Node<E> node) {
        while ((node = node.parent) != null) {
            if (isBalanced(node)) {
                updateHeight(node);
            }else {
                rebalance(node);
            }
        }
    }

    private void rebalance(Node<E> grand) {
        Node<E> parent = ((AVLNode<E>)grand).tallerChild();
        Node<E> node = ((AVLNode<E>)parent).tallerChild();

        if (parent.isLeftChild()) {
            if (node.isLeftChild()){ // LL
                rotate(grand, node.left, node, node.right, parent, parent.right, grand, grand.right);
            }else { //LR
                rotate(grand, parent.left, parent, node.left, node, node.right, grand, grand.right);
            }
        }else {
            if (node.isLeftChild()){ // RL
                rotate(grand, grand.left, grand, node.left, node, node.right, parent, parent.right);
            }else { //RR
                rotate(grand, grand.left, grand, parent.left, parent, node.left, node, node.right);
            }
        }
    }

    @Override
    protected void afterRotate(Node<E> grand, Node<E> parent, Node<E> child) {
        super.afterRotate(grand, parent, child);

        updateHeight(grand);
        updateHeight(parent);
    }

    @Override
    protected void rotate(Node<E> r, Node<E> a, Node<E> b, Node<E> c, Node<E> d, Node<E> e, Node<E> f, Node<E> g) {
        super.rotate(r, a, b, c, d, e, f, g);

        updateHeight(b);
        updateHeight(f);
        updateHeight(d);
    }

    private void rebalanceIndp(Node<E> grand) {

        Node<E> parent = ((AVLNode<E>)grand).tallerChild();
        Node<E> node = ((AVLNode<E>)parent).tallerChild();

        if (parent.isLeftChild()) {
            if (node.isLeftChild()){ // LL
                rotateRight(grand);
            }else { //LR
                rotateLeft(parent);
                rotateRight(grand);
            }
        }else {
            if (node.isLeftChild()){ // RL
                rotateRight(parent);
                rotateLeft(grand);
            }else { //RR
                rotateLeft(grand);
            }
        }
    }


    private void updateHeight(Node<E> node) {
        ((AVLNode)node).updateHeight();
    }

    @Override
    protected Node<E> createNode(E element, Node<E> parent) {
        return new AVLNode<E>(element, parent);
    }

    private boolean isBalanced(Node<E> node) {
        return Math.abs(((AVLNode<E>)node).balanceFactor()) <= 1;
    }

    private static class AVLNode<E> extends Node<E>{
        int height = 1;

        public AVLNode(E element, Node<E> parent){
            super(element, parent);
        }

        public int balanceFactor() {
            int leftHeight = left == null ? 0 : ((AVLNode)left).height;
            int rightHeight = right == null ? 0 : ((AVLNode)right).height;
            return leftHeight - rightHeight;
        }

        public void updateHeight() {
            int leftHeight = left == null ? 0 : ((AVLNode)left).height;
            int rightHeight = right == null ? 0 : ((AVLNode)right).height;
            height = 1 + Math.max(leftHeight, rightHeight);
        }


        public Node<E> tallerChild() {
            int leftHeight = left == null ? 0 : ((AVLNode)left).height;
            int rightHeight = right == null ? 0 : ((AVLNode)right).height;
            if (leftHeight > rightHeight) return left;
            if (leftHeight < rightHeight) return right;
            return isLeftChild() ? left : right;
        }

        @Override
        public String toString() {
            String parentString = "null";
            if (parent != null) {
                parentString = parent.element.toString();
            }
            return element + "_h(" + height + ")";
        }
    }
}

File: Python/Tree/avltree.py
## written by xiongbiao
## date 2020-5-20


from Tree.bbstree import BalancedBinarySearchTree




class AVLTree(BalancedBinarySearchTree):
    class AVLNode(BalancedBinarySearchTree.Node):
        def __init__(self, value, parent):
            super().__init__(value, parent)
            self._height = 1

        '''
        AVL树节点方法：更新节点的高度，用于平衡后重新计算节点的高度
        '''

        def update_height(self):
            if self.left:
                left_height = self.left._height
            else:
                left_height = 0

            if self.right:
                right_height = self.right._height
            else:
                right_height = 0

            self._height = 1 + max(left_height, right_height)

        '''
        AVL树节点方法：计算节点的平衡因子，计算当前节点的左右子树的高度差
        '''

        def balance_factor(self):
            if self.left:
                left_height = self.left._height
            else:
                left_height = 0

            if self.right:
                right_height = self.right._height
            else:
                right_height = 0
            return left_height - right_height

        '''
        AVL树节点方法：获取左右子节点高度高的那棵子树，用于AVL树添加删除后旋转中提取不平衡的子树的节点，
                     如果2个子树高度一样，返回同方向节点
        '''

        def taller_child(self):
            if self.left:
                left_height = self.left._height
            else:
                left_height = 0

            if self.right:
                right_height = self.right._height
            else:
                right_height = 0

            if left_height > right_height:
                return self.left
            if left_height < right_height:
                return self.right

            if self.is_left_child():
                return self.left
            else:
                return self.right

    def __init__(self):
        super().__init__()
    '''
    创建AVL节点
    '''
    def create_node(self, value, parent):
        return AVLTree.AVLNode(value, parent)

    '''
    添加节点后，顺着节点的父节点向上找，若节点平衡更新节点高度，找到第一个不平衡的节点，平衡该节点，结束
    '''
    def after_add(self, node):
        node = node.parent
        while node:
            if self.is_balance(node):
                self.update_height(node)
            else:
                self.rebalance(node)
                break
            node = node.parent

    '''
    删除节点后，顺着节点的父节点向上找，若节点平衡更新节点高度，找到不平衡的节点，平衡该节点，直至root
    '''
    def after_remove(self, node):
        node = node.parent
        while node:
            if self.is_balance(node):
                self.update_height(node)
            else:
                self.rebalance(node)
            node = node.parent

    '''
    判断节点是否平衡，节点平衡因子绝对值<=1
    '''
    def is_balance(self, node):
        return abs(node.balance_factor()) <= 1

    '''
    更新节点高度
    '''
    def update_height(self, node):
        node.update_height()

    '''
    平衡节点，找到节点的子节点中高的子节点和孙子节点中高的子节点，判断子节点与孙子节点的方向，旋转相应的节点
    1. LL：parent是grand的left，child是parent的left，右旋grand
    2. LR：parent是grand的left，child是parent的right，左旋parent，变成LL，右旋grand
    3. RL：parent是grand的right，child是parent的left，右旋parent，变成RR，，左旋grand
    4. RR：parent是grand的right，child是parent的right，左旋grand
    '''
    def rebalance(self, grand):
        parent = grand.taller_child()
        child = parent.taller_child()

        if parent.is_left_child():
            if child.is_left_child():
                self.rotate_right(grand)
            else:
                self.rotate_left(parent)
                self.rotate_right(grand)
        else:
            if child.is_left_child():
                self.rotate_right(parent)
                self.rotate_left(grand)
            else:
                self.rotate_left(grand)

    def after_rotate(self, grand, parent, child):
        super().after_rotate(grand, parent, child)
        self.update_height(grand)
        self.update_height(parent)

# nodes = [67, 81, 77, 60, 11, 87, 89, 84, 69]
# tree = AVLTree()
# for n in nodes:
#     tree.add_value(n)
#
# tree.remove_value(11)
# tree.remove_value(60)
# tree.remove_value(67)
# tree.print()